apps/dhparam.c

```diff
@@ -313,29 +313,29 @@ int dhparam_main(int argc, char **argv)
                         "# include <openssl/dh.h>\n"
                         "#endif\n"
                         "\n");
-        BIO_printf(out, "DH *get_dh%d()\n{\n", bits);
+        BIO_printf(out, "static DH *get_dh%d()\n{\n", bits);
         print_bignum_var(out, pbn, "dhp", bits, data);
         print_bignum_var(out, gbn, "dhg", bits, data);
         BIO_printf(out, "    DH *dh = DH_new();\n"
-                        "    BIGNUM *dhp_bn, *dhg_bn;\n"
+                        "    BIGNUM *p, *g;\n"
                         "\n"
                         "    if (dh == NULL)\n"
                         "        return NULL;\n");
-        BIO_printf(out, "    dhp_bn = BN_bin2bn(dhp_%d, sizeof(dhp_%d), NULL);\n",
+        BIO_printf(out, "    if (!DH_set0_pqg(dh,\n");
```

Comment:
This will break.  The `DH_set0_pqg` function only permits **q** to be `NULL`.  It doesn't make any such claim about **p** or **g**.

Reply:
That's the point. If for any reason the `BN_bin2bn()` fail, we want the `DH_set0_pqg()` to fail as well.

Reply:
According to the documentation, `DH_set0_pqg` can do anything if **p** or **g** is `NULL`.  It won't necessarily fail, it could crash or it could immolate the programmer.  The current code catches `NULL` but these checks are being (slowly) removed as per the current policy.

Reply:
> The current code catches NULL but these checks are being (slowly) removed as per the current policy.

??? Goal is not to remove all NULL checks, but to remove **redundant** checks. Those that can't be NULL (e.g. EC_GROUP->order), or are not permitted to be NULL (e.g. similar to memcpy).

As for break or not. One has to recognize that we are looking at *new* structure, i.e. without parameters being set. In this case set0 performs the NULL checks. And those checks are not going away. Well, one can still argue that this is equivalent to making an assumption about implementation...

Reply:
> And those checks are not going away.

This is naturally an overstatement in sense that it's not like it's up to me :-) It rather means that I for one don't see that removal of referred NULL checks would be justified. And this deserves clarification. As long as DH_set0_pqg permits setting them separately, e.g. as (p, NULL, NULL) followed by (NULL,NULL,g), NULL checks will have to be there. But even if we demand that all parameters are passed in single call, I'd still argue in favour of keeping initial check as it is. Because it should allow to make checks elsewhere redundant and "omitable". Basically it's all about balance. You have to have *some* checks, and question is where (as close to user as possible) and how many are they (as little as possible).